basics/units.ipynb

```diff
@@ -0,0 +1,403 @@
+    }
+   ],
+   "source": [
+    "GeV == 1000*MeV"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Add two quantities with different length units:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "1050.0"
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "from hepunits import units as u\n",
+    "\n",
+    "1*u.meter + 5*u.cm"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The result is in HEP units, so mm. Indeed, for example `u.meter == 1000 == 1000 * u.mm`. \n",
+    "\n",
+    "Rather obtain the result in meters:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "1.05"
+      ]
+     },
+     "execution_count": 4,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "(1*u.meter + 5*u.cm) / u.meter"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Do you need to play a bit more to get a proper feeling? This next (non-academic) exercise should help you ..."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### **Quick time-of-flight study**\n",
+    "\n",
+    "Let's try to play with units in a meaningful way, in a kind of exercise that physicists encounter. Imagine you are investigating time-of-flight (ToF) detectors for particle identification. The time it takes a particle of velocity $\\beta = v/c= pc/E$ to travel a distance $L$ is given by\n",
+    "\n",
+    "$$\\mathrm{ToF} = \\frac{L}{c \\beta}$$\n",
+    "\n",
+    "It results that the mass $m$ of the particle can be determined from\n",
+    "\n",
+    "$$m = \\frac{p}{c}\\sqrt{\\frac{c^2 \\mathrm{ToF}^2}{L^2}-1}$$\n",
+    "\n",
+    "provided the path length $L$ and the momentum $p$ can be measured, say, by a tracking system."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "What are typical ToF differences say for (charged) kaons and pions?\n",
+    "It is practical to perform the calculation as\n",
+    "\n",
+    "$$\\Delta \\mathrm{ToF} = \\frac{L}{c}(\\frac{1}{\\beta_1} - \\frac{1}{\\beta_2})\\,,$$\n",
+    "\n",
+    "with $\\frac{1}{\\beta} = \\sqrt{1+m^2c^2/p^2}$."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from hepunits import c_light, GeV, meter, ps, ns\n",
+    "\n",
+    "import numpy as np\n",
+    "\n",
+    "def ToF(m, p, L):\n",
+    "    \"\"\"Time-of-Flight = particle path length L / (c * beta)\"\"\"\n",
+    "    one_over_beta = np.sqrt(1 + m*m/(p*p))  # no c factors here because m and p given without them, hence c's cancel out ;-)\n",
```

Comment:
I am not sure whether this is a good example to demonstrate the unit system, since it is a bit of a sloppy (although very highly practical and every-day) use of units in HEP. On a strictly superficial level one could conclude from this that units are not so important, they rather get in the way and being precise about units is not practical. I suppose that is not really the message you want to convey with units.

Since the units do not add additional error checking, maybe worthwhile to point out that they add a lot to readability of the code (although correctness must be still manually checked).

Reply:
Indeed readability is a big plus at least to me since `5 * GeV` tells me something more than just `5`, for the sake of argument. I will add a statement on this.

Yes I also realised that some bits of code are not straightforward. My idea was to show how ti use the package for a (potentially) real-life example, and I in fact did the "study" myself preparing the notebook, as I had not thought about ToF before and what are the kind of numbers / time resolutions one typically talks about. It was useful to me :-).

Reply:
I like the example as it "makes me forget" about the units (scaling), the orders of magnitudes. A distance can be given in meters or whatever and it matches. The error checking and auto conversion would be a completely different level of package, so it seems to fit nicely, as this is what the package should do and that's also it. For it's striking simplicity, you get a lot out of it.
@HDembinski what else would you have in mind as an example, which direction?

Reply:
To clarify, my suggestion is to use another example which does not deal with the subtlety that we often use energies in place of momenta and masses, which allows us to write our formulas without factors of c. I think this a subtlety that is distracting from the main message here, that using units **consistently** is important.

The fact that you can even do this (pass an energy in place of a momentum) is a violation of the unit system, unless you explicitly use a system in which c is defined to be 1 and then energies, masses, and momenta have the same base unit, likewise lengths and times.

Reply:
If you want to keep the example, then I would at least refer to the point that this unit system is intentionally not actually checking units, so you can use units in this lax way.

Reply:
That's a good point indeed, I misunderstood in the first place. In fact, an example with correct units, even if maybe simple, could be useful. Thinking about it...